From 0007e701fb8583599b645cb6e8fbd9a9481f9b14 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pablo=20Correa=20G=C3=B3mez?= <ablocorrea@hotmail.com>
Date: Wed, 18 Sep 2024 20:12:35 +0200
Subject: [PATCH 3/4] plugins: appstream: abort refine_from_id if there are no
 components

Else, found will be set unconditionally to TRUE, and the calling
function, refine_thread_cb, will not go into refine_from_pkgname,
despite nothing having been found.

We have identified this during testing of an OOT plugin.

Fixes 5dd59c4a4752fab9e5e83c498dfc32a33263edba
---
 plugins/core/gs-plugin-appstream.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/plugins/core/gs-plugin-appstream.c b/plugins/core/gs-plugin-appstream.c
index 23e4ab92b..aa39f94d2 100644
--- a/plugins/core/gs-plugin-appstream.c
+++ b/plugins/core/gs-plugin-appstream.c
@@ -1050,14 +1050,16 @@ gs_plugin_refine_from_id (GsPluginAppstream    *self,
 	} else {
 		components = g_hash_table_lookup (apps_by_id, id);
 	}
-	if (components != NULL) {
-		for (guint i = 0; i < components->len; i++) {
-			XbNode *component = g_ptr_array_index (components, i);
-			if (!gs_appstream_refine_app (GS_PLUGIN (self), app, self->silo, component, flags, self->silo_installed_by_desktopid,
-						      self->silo_filename ? self->silo_filename : "", self->default_scope, error))
-				return FALSE;
-			gs_plugin_appstream_set_compulsory_quirk (app, component);
-		}
+
+	if (components == NULL)
+		return TRUE;
+
+	for (guint i = 0; i < components->len; i++) {
+		XbNode *component = g_ptr_array_index (components, i);
+		if (!gs_appstream_refine_app (GS_PLUGIN (self), app, self->silo, component, flags, self->silo_installed_by_desktopid,
+					      self->silo_filename ? self->silo_filename : "", self->default_scope, error))
+			return FALSE;
+		gs_plugin_appstream_set_compulsory_quirk (app, component);
 	}
 
 	/* if an installed desktop or appdata file exists set to installed */
-- 
2.46.1

